test/test_ff_mocks.py

```diff
@@ -1,10 +1,516 @@
+
+    # Test normal result
+    stream = io.StringIO()
+    r.recording_fp = stream
+    PRINT(json.dumps({'verb': 'GET', 'url': "http://foo", 'data': None, 'duration': 10.0, 'status': 200,
+                      'result': {"some": "result"}}),
+          file=stream)
+    stream.seek(0)
+
+    response = r.do_mocked_replay(verb='GET', url="http://foo")
+    assert response.status_code == 200
+    assert response.json() == {"some": "result"}
+
+    # Test error result
+    stream = io.StringIO()
+    r.recording_fp = stream
+    PRINT(json.dumps({'verb': 'GET', 'url': "http://foo", 'data': None, 'duration': 10.0, 'status': 200,
+                      'error_type': 'RuntimeError', 'error_message': 'yikes'}),
+          file=stream)
+    stream.seek(0)
+
+    with pytest.raises(Exception) as exc:
+        r.do_mocked_replay(verb='GET', url="http://foo")
+    assert not isinstance(exc, AssertionError)
+    assert str(exc.value) == "yikes"
+
+    # Test mismatched recording call
+    stream = io.StringIO()
+    r.recording_fp = stream
+    PRINT(json.dumps({'verb': 'POST', 'url': 'http://any', 'data': None, 'duration': 17.0,
+                      'result': 'ignored-result'}),
+          file=stream)
+    stream.seek(0)
+
+    with pytest.raises(AssertionError) as exc:
+        r.do_mocked_replay(verb='GET', url="http://foo")
+    assert str(exc.value) == "Actual call GET http://foo does not match expected call POST http://any"
+
+    # Test mismatched recording data
+    stream = io.StringIO()
+    r.recording_fp = stream
+    PRINT(json.dumps({'verb': 'POST', 'url': 'http://foo', 'data': 'something', 'duration': 17.0,
+                      'result': 'ignored-result'}),
+          file=stream)
+    stream.seek(0)
+
+    with pytest.raises(AssertionError) as exc:
+        r.do_mocked_replay(verb='POST', url="http://foo", data='something-else')
+    assert str(exc.value) == "Data mismatch on call POST http://foo."
+
+
+@using_fresh_legacy_state_for_testing()
+def test_test_scenario_methods():
+
+    assert TestScenarios.mocked_auth_key('foo') == 'fookey'
+    assert TestScenarios.mocked_auth_key('fourfront-foo') == 'fookey'
+    assert TestScenarios.mocked_auth_key('fourfront-cgapfoo') == 'cgapfookey'
+    assert TestScenarios.mocked_auth_key('cgap-foo') == 'cgap-fookey'
+
+    assert TestScenarios.mocked_auth_secret('foo') == 'foosecret'
+    assert TestScenarios.mocked_auth_secret('fourfront-foo') == 'foosecret'
+    assert TestScenarios.mocked_auth_secret('fourfront-cgapfoo') == 'cgapfoosecret'
+    assert TestScenarios.mocked_auth_secret('cgap-foo') == 'cgap-foosecret'
+
+    assert TestScenarios.mocked_auth_server('foo') == 'http://foo.4dnucleome.org/'
+    assert TestScenarios.mocked_auth_server('fourfront-foo') == 'http://foo.4dnucleome.org/'
+    assert TestScenarios.mocked_auth_server('fourfront-cgapfoo') == 'http://cgapfoo.4dnucleome.org/'
+    assert TestScenarios.mocked_auth_server('cgap-foo') == 'http://cgap-foo.4dnucleome.org/'
```

Comment:
Do we always want a .4dnucleome domain?

Reply:
These are test scenarios. It's more a matter of running through the code path. I could just match on a regexp that it returned something like `http://cgap-foo[.].*[.](com|org)/`.

Reply:
I looked at it and think I’ll leave it for now because there are too many cases and the regexp would complicate things. It’s one of those cases where the test is just holding the coverage in check, but also saying “this is what the example returns” not “this is what code has to do”. if someone changes that TestScnaerio it will have to change the test, but that’s not horrible. They don’t change often.